src/System.Linq/src/System/Linq/Partition.cs

```diff
@@ -389,5 +389,310 @@ public int GetCount(bool onlyIfCheap)
+            {
+                if (onlyIfCheap)
+                {
+                    return -1;
+                }
+
+                if (!HasLimit)
+                {
+                    // If HasLimit is false, we contain everything past _minIndexInclusive.
+                    // Therefore, we have to iterate the whole enumerable.
+                    return Math.Max(_source.Count() - _minIndexInclusive, 0);
+                }
+
+                using (IEnumerator<TSource> en = _source.GetEnumerator())
+                {
+                    // We only want to iterate up to _maxIndexInclusive + 1.
+                    // Past that, we know the enumerable will be able to fit this partition,
+                    // so the count will just be _maxIndexInclusive + 1 - _minIndexInclusive.
+
+                    // Note that it is possible for _maxIndexInclusive to be int.MaxValue here,
+                    // so + 1 may result in signed integer overflow. We need to handle this.
+                    // At the same time, however, we are guaranteed that our max count can fit
+                    // in an int because if that is true, then _minIndexInclusive must > 0.
+
+                    uint count = SkipAndCount((uint)_maxIndexInclusive + 1, en);
+                    Debug.Assert(count != (uint)int.MaxValue + 1 || _minIndexInclusive > 0, "Our return value will be incorrect.");
+                    return Math.Max((int)count - _minIndexInclusive, 0);
+                }
+
+            }
+
+            public override bool MoveNext()
+            {
+                // Cases where GetEnumerator has not been called or Dispose has already
+                // been called need to be handled explicitly, due to the default: clause.
+                int taken = _state - 3;
+                if (taken < -2)
+                {
+                    Dispose();
+                    return false;
+                }
+
+                switch (_state)
+                {
+                    case 1:
+                        _enumerator = _source.GetEnumerator();
+                        _state = 2;
+                        goto case 2;
+                    case 2:
+                        if (!SkipBeforeFirst(_enumerator))
+                        {
+                            // Reached the end before we finished skipping.
+                            break;
+                        }
+
+                        _state = 3;
+                        goto default;
+                    default:
+                        if ((!HasLimit || taken < Limit) && _enumerator.MoveNext())
+                        {
+                            if (HasLimit)
+                            {
+                                // If we are taking an unknown number of elements, it's important not to increment _state.
+                                // _state - 3 may eventually end up overflowing & we'll hit the Dispose branch even though
```

Comment:
> Would it be worth doing something like that here? Or we don't think there's enough risk to warrant that kind of fragility?

It is low risk, but at the same time I think it is unlikely the field name of `_state` will change. I will add a test.